File: data-structures/linked-list/doubly-linked-list/DoublyLinkedList.java
class DoubleyLinkedList {
  Node head, tail;

  class Node {
    int data;
    Node prev;
    Node next;

    Node(int data) {
      this.data = data;
      this.prev = null;
      this.next = null;
    }
  }

  public DoubleyLinkedList() {
    head = null;
    tail = null;
  }

  // Insert at front
  void insertFirst(int data) {
    Node newNode = new Node(data);

    if (head == null) {
      head = tail = newNode;
    } else {
      newNode.next = head;
      head.prev = newNode;
      head = newNode;
    }
  }

  // Insert at end
  void insertLast(int data) {
    Node newNode = new Node(data);

    if (tail == null) {
      head = tail = newNode;
    } else {
      tail.next = newNode;
      newNode.prev = tail;
      tail = newNode;
    }
  }

  // Delete from front
  void deleteFirst() {
    if (head == null) {
      System.out.println("Lsit is Empty");
      return;
    }

    if (head == tail) {
      head = tail = null;
    } else {
      head = head.next;
      head.prev = null;
    }
  }

  // Delete fron end
  void deleteLast() {
    if (tail == null) {
      System.out.println("List is Empty");
      return;
    }
    if (head == tail) {
      head = tail = null;
    } else {
      tail = tail.prev;
      tail.next = null;
    }
  }

  // Display from front
  void displayForward() {
    Node current = head;
    System.out.print("Forward: ");
    while (current != null) {
      System.out.print(current.data + " ");
      current = current.next;
    }
    System.out.println();
  }

  // Display from end
  void displayBackward() {
    Node current = tail;
    System.out.print("Backward: ");
    while (current != null) {
      System.out.print(current.data + " ");
      current = current.prev;
    }
    System.out.println();
  }
}

File: data-structures/linked-list/doubly-linked-list/DoublyLinkedList.cpp
#include <iostream>
using namespace std;

class Node {
public:
  int data;
  Node *next;
  Node *prev;

  Node(int value) {
    data = value;
    next = prev = nullptr;
  }
};

class DoublyLinkedList {
  Node *head;
  Node *tail;

public:
  DoublyLinkedList() { head = tail = nullptr; }

  void push_front(int value) {
    Node *newNode = new Node(value);

    if (head == nullptr) {
      head = tail = newNode;
      return;
    }

    newNode->next = head;
    head->prev = newNode;
    head = newNode;
  }

  void push_back(int value) {
    Node *newNode = new Node(value);

    if (head == nullptr) {
      head = tail = newNode;
      return;
    }

    newNode->prev = tail;
    tail->next = newNode;
    tail = newNode;
  }

  void pop_front() {
    if (head == nullptr) {
      cout << "Doubly Linked List is empty!\n";
      return;
    }

    Node *temp = head;

    head = head->next;

    if (head != nullptr) {
      head->prev = nullptr;
    }

    temp->next = nullptr;
    delete temp;

    if (head == nullptr) {
      tail = nullptr;
    }
  }

  void pop_back() {
    if (head == nullptr) {
      cout << "Doubly Linked List is empty!\n";
      return;
    }

    if (head == tail) {
      delete head;
      head = tail = nullptr;
      return;
    }

    Node *temp = tail;
    
    tail = tail->prev;

    if (tail != nullptr) {
      tail->next = nullptr;
    }

    temp->prev = nullptr;
    delete temp;
  }

  void print() {
    Node *temp = head;

    while (temp != nullptr) {
      cout << temp->data << " <-> ";
      temp = temp->next;
    }

    cout << "NULL\n";
  }
};
